Combine ip and ip2 libraries

kgerheiser opened the pull request:
I've been working on this for the past two weeks and I finally have a combined ip library. All the grib1/grib2 tests pass in this one repo. This is a Draft PR while I add some final touches. The history is based of off ip2, but I plan to make the actual PR into NCEPLIBS-ip.

Instead of operating on the raw grib1/grib2 descriptors the code has been refactored to make use of some new objects.

To start with there's the abstract type `ip_grid_descriptor` in `ip_grid_descriptor_mod.f90`, and there's two subclasses `grib1_descriptor` and `grib2_descriptor` which basically just wrap the grib1/grib2 `kgds` and `igdtmpl` arrays. To create a descriptor you call `init_descriptor` and pass the grib arrays.

Then, there are the grid classes which are created by calling `init_grid` with an `ip_grid_descriptor` object. At the top-level there's the abstract type `ip_grid` which contains common variables for each grid like `im`, `jm`, `rerth`, etc. Each concrete subclass of `ip_grid` (e.g. `ip_equid_cylind_grid`) contains grid-specific variables and must implement the routines `init_grib1`, `init_grib2` and `gdswzd`. The init routines are responsible for handling grib1/grib2 specific  initializations and setting the proper variables.

The top-level `ipolates(v)` routines retain their interfaces and then create descriptors, which are used to create the grid object, which is then passed to the actual interpolation routines. 

With that framework, all interpolation routines and gdswzd routines now take grid objects instead of passing the descriptors around.


Spectral interpolation has not been merged together yet, and I just copied the grib1/grib2 routines into the module with an interface. There are a lot of specific checks using the grib descriptor arrays that are difficult to generalize. I'm still looking into this.


I am working on cleaning up the code, adding documentation, and doing some final touches.

kgerheiser committed:
Merge grib1 tests

kgerheiser committed:
Create ip_mod as top-level module and wrap ipolates into a module

kgerheiser committed:
Indent module

kgerheiser committed:
Put polates(v) into modules

kgerheiser committed:
Add grib1 gdswzd modules

kgerheiser committed:
Put ipolates and ipolatev into modules

kgerheiser committed:
Put ijkgds0 and ijkgds1 into a module

kgerheiser committed:
Remove ip2 suffix from gdswzd_mod

kgerheiser committed:
Update list-of-files with grib1 gdswzd

kgerheiser committed:
Initial merge of ip and ip2. All tests pass

kgerheiser committed:
Add gdswzd routine for grid object

kgerheiser committed:
Bring in specific grid types and give them gdswzd type-bound procedure

kgerheiser committed:
Disable building 4/8 version of library while testing

kgerheiser committed:
Use grid routines for gdswzd and ijkgds1 to start full transition

kgerheiser committed:
Revert grib1 polates0

kgerheiser committed:
Add verbose to test

kgerheiser committed:
Rlon is not necessary to pass to polfixs

kgerheiser committed:
Put polfix routines into a module

kgerheiser committed:
polates0 grib1 use grid

kgerheiser committed:
Use grid object for polates1

kgerheiser committed:
Add debug flags

kgerheiser committed:
Revert grid object in polates1

kgerheiser committed:
Disable grid object

kgerheiser committed:
Use grid object in polates0

kgerheiser committed:
Change gdswzd

kgerheiser committed:
Use old gdswzd

kgerheiser committed:
Use old ijkgds1

kgerheiser committed:
Change gdswzd to use grid

kgerheiser committed:
Update nscan_field_pos for ijkgds1

kgerheiser committed:
Use grid%field_pos now that nscan has been updated properly

kgerheiser committed:
Convert polates1 to use grid object

kgerheiser committed:
Convert polates2 to use grid object

kgerheiser committed:
Use grid object in polates3

kgerheiser committed:
Use grid object for polates4

kgerheiser committed:
Use grid object for polates6

kgerheiser committed:
Add ip_grid to ip_grids_mod

kgerheiser committed:
Convert polates0 to only use grid

kgerheiser committed:
Merge polates1 grib1 and grib2

kgerheiser committed:
Remove polates1 grib1 and grib2 specific routines

kgerheiser committed:
Merge grib1 and grib2 polates2

kgerheiser committed:
Remove grib1 and grib2 specifc routines from polates2

kgerheiser committed:
Merge grib1 and grib2 polates3

kgerheiser committed:
Remove grib1 and grib2 specific routines for polates3

kgerheiser committed:
Merge grib1 and grib2 routines for polates6

kgerheiser committed:
Remove grib1 and grib2 specific polates6

kgerheiser committed:
Make grib1 and grib2 specific gdswzd routines use grid objects

kgerheiser committed:
Enable gdswzd_c_grib2 test

kgerheiser committed:
Convert vector bilinear interp to grid object

kgerheiser committed:
Convert polatev1 to grid object

kgerheiser committed:
Convert polatev2 to grid object

kgerheiser committed:
Convert polatev3 to grid object

kgerheiser committed:
Convert polatev6 to use grid objects

kgerheiser committed:
Cleanup code by removing unused modules, functions, variables, etc

kgerheiser committed:
Use grid object for gdswzd in spectral interpolation

kgerheiser committed:
Remove unused files

GeorgeGayno-NOAA commented:
@kgerheiser When you are ready, please compile it on one of our officially supported machines.  I would like to test it with UFS_UTILS.

kgerheiser committed:
Re-name interp modules with actual name instead of number

kgerheiser committed:
Put vector and scalar interpolation routines into same module

kgerheiser committed:
Create ipolatev_grid

kgerheiser committed:
Change project name to ip

kgerheiser committed:
Add gdswzd to ip_mod

kgerheiser commented:
@GeorgeGayno-NOAA 

I installed in on Hera.

Set `export ip_ROOT=/home/Kyle.Gerheiser/NCEPLIBS-ip2/build/install` to use it with CMake in UFS_UTILS. 

You'll have to add `use ip_mod` wherever you use the library. That module contains all the public facing routines: `ipolates`, `ipolatev`, and `gdswzd`.

It's possible you might also have to adjust your calls because now that everything is in a module you have to match the interface exactly. Like, in the test code `ibi` was being passed as a scalar when the interface says it's supposed to be an array. So, I changed it to an array of size one. Or `rlat/rlon` were declared as 2d arrays, but the `polate` routines take 1-d arrays. Don't know if you do anything like that.


And the C routines were changed. C doesn't have function overloading so the grib1/grib2 C versions of `gdswzd` can't both  be called `gdswzd`. I left the `ip2` version alone and named the `ip1` version `gdswzd_grib1`.

kgerheiser commented:
Building UFS_UTILS I ran into what I thought might happen. In `snowdat.f` in UFS_UTILS `rlat` and `rlon` are 2-d arrays, but the interface wants a 1-d array. I think this is because of the `interface gdswzd` and it can't find a matching interface that takes 2-d arrays.

I think this can be fixed by adding a `gdswzd_2d_array_grib1` in the same way there's a grib2 `gdswzd_2d_array` in ip2

kgerheiser commented:
I was able to get UFS_UTILS to build. You can the changes I had to make:

https://github.com/kgerheiser/UFS_UTILS/commit/a7a8d84879eeec850dafaebc7f295de6c38b893b

There's a lot of inconsistency with the shape of various arrays (sometimes 1-d, sometimes 2-d). Using an interface is a no-go because of that, and I didn't want to change anything. So, I used the specific call directly `ipolates_grib1` because that allows you to pass 1-d arrays to 2-d arguments or vice-versa.

I also had to change some `intent(in)` because in `ip` they're `intent(inout)` and you can't pass an `intent(in)` argument to a routine that's `intent(inout)`.

GeorgeGayno-NOAA commented:
> I was able to get UFS_UTILS to build. You can the changes I had to make:
> 
> [kgerheiser/UFS_UTILS@a7a8d84](https://github.com/kgerheiser/UFS_UTILS/commit/a7a8d84879eeec850dafaebc7f295de6c38b893b)
> 
> There's a lot of inconsistency with the shape of various arrays (sometimes 1-d, sometimes 2-d). Using an interface is a no-go because of that, and I didn't want to change anything. So, I used the specific call directly `ipolates_grib1` because that allows you to pass 1-d arrays to 2-d arguments or vice-versa.
> 
> I also had to change some `intent(in)` because in `ip` they're `intent(inout)` and you can't pass an `intent(in)` argument to a routine that's `intent(inout)`.

Thanks.  I will open an issue and create a branch with your updates.

kgerheiser committed:
Build '4' version of library

kgerheiser committed:
Create 2d version of grib1 gdswzd

kgerheiser committed:
Make specific interpolation routines public

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA
> 
> I installed in on Hera.
> 
> Set `export ip_ROOT=/home/Kyle.Gerheiser/NCEPLIBS-ip2/build/install` to use it with CMake in UFS_UTILS.
> 
> You'll have to add `use ip_mod` wherever you use the library. That module contains all the public facing routines: `ipolates`, `ipolatev`, and `gdswzd`.
> 
> It's possible you might also have to adjust your calls because now that everything is in a module you have to match the interface exactly. Like, in the test code `ibi` was being passed as a scalar when the interface says it's supposed to be an array. So, I changed it to an array of size one. Or `rlat/rlon` were declared as 2d arrays, but the `polate` routines take 1-d arrays. Don't know if you do anything like that.
> 
> And the C routines were changed. C doesn't have function overloading so the grib1/grib2 C versions of `gdswzd` can't both be called `gdswzd`. I left the `ip2` version alone and named the `ip1` version `gdswzd_grib1`.

@kgerheiser Can you open up the permissions on your build directory?

kgerheiser commented:
Try it now

GeorgeGayno-NOAA commented:
The UFS_UTILS snow2mdl regression test is failing with a seg fault.   It happens in routine snowdat.f at line 1257.  I tried two ways to call gdswzd.

`  call gdswzd_grib1(kgds,(-1),npts,fill,xpts,ypts,rlon,rlat,nret)` and
` call gdswzd(kgds,(-1),npts,fill,xpts,ypts,rlon,rlat,nret)`

In both cases, the routine returns nret=0.  And xpts(1) and ypts(1) = 9999.   The input arrays, kgds, rlat, rlon are correct.  You can see the log files on Hera: /scratch2/NCEPDEV/stmp1/George.Gayno/reg_tests.snow2mdl.seg.fault.

kgerheiser commented:
What kind of grid is it running on?

How can I run just that test and do it interactively?

kgerheiser commented:
I think I have it working. All I need is that `fort.41` file? I turned on debug flags and see an out-of-bounds error.

kgerheiser commented:
I think I found the bug (or at least a bug):

In `gdswzd_05` there's this line:
`IF(H.EQ.-1)ORIENT=ORIENT+180.`

But in ip2 there isn't, and I guess I missed it when I was comparing the two files.

kgerheiser committed:
Fix ORIENT calculation in polar stereo grid when H == -1

kgerheiser commented:
I ran the tests myself, and it still doesn't pass, but I plotted it and compared it to the original, and it's close. Here is the relative error between them (array2 - array1) / array1

<img width="1017" alt="Screen Shot 2021-01-21 at 14 21 58" src="https://user-images.githubusercontent.com/3209794/105401292-080ab780-5bf4-11eb-8ac3-12a3dc8840d2.png">

I'll see if I can find what the issue is.

kgerheiser commented:
I am going to use the base UFS_UTILS and change one call at a time while testing at each stage and link in both the new library and the old library to track where the difference comes from.

kgerheiser committed:
Change project name to test with UFS_UTILS

kgerheiser committed:
Revert project name back to ip

kgerheiser committed:
Use correct earth radius with elliptic grid

GeorgeGayno-NOAA commented:
@kgerheiser I will try your fix at 9aab5fe.

edwardhartnett commented on src/constants_mod.f90:
```diff
@@ -8,5 +8,8 @@ module constants_mod
   real(real64), parameter :: pi2=pi/2.0d0
   real(real64), parameter :: pi4=pi/4.0d0
+  real(real64), parameter :: RERTH_WGS84=6.378137E6
+  real(real64), parameter :: E2_WGS84 = 0.00669437999013
```
Do we now have a test that requires these parameters be correct?

edwardhartnett commented:
Every time you make a fix, you should also ensure there is a test that will fail without that fix.

If you discover some problems with external, manual testing, then you must update the internal, automatic tests, so that they will find the problem next time.

Although it might seem that every bug is unique, careful study has shown that programmers make the same few mistakes over and over again. So putting a test in place every time that happens is a great way to build up a set of test which will be valuable in the future, catching the mistakes immediately.

kgerheiser commented:
I can add a new test that uses that particular grid. There's a polar stereo grid test, but it doesn't test the elliptical part.

@GeorgeGayno-NOAA would I be able to edit the output descriptor of the current `lat-lon -> polar stereo grid` test to and just make it elliptical? Then, I could test it against the existing ip and ip2 library with the same settings.

GeorgeGayno-NOAA commented:
> I can add a new test that uses that particular grid. There's a polar stereo grid test, but it doesn't test the elliptical part.
> 
> @GeorgeGayno-NOAA would I be able to edit the output descriptor of the current `lat-lon -> polar stereo grid` test to and just make it elliptical? Then, I could test it against the existing ip and ip2 library with the same settings.

@kgerheiser Point me to the source code file for the `lat-lon to polar grid` test.

kgerheiser commented:
Sure, it's the tests you created, but very slightly modified and combined with CMake:

https://github.com/kgerheiser/NCEPLIBS-ip2/blob/feature/combined-ip/tests/interp_mod_grib1.f90

This descriptor specifically (there's also `interp_mod_grib2.f90`):

https://github.com/kgerheiser/NCEPLIBS-ip2/blob/9aab5fefb03a8dd8994ef8e03ea2107b95cd5706/tests/interp_mod_grib1.f90#L85

Here is where the top-level executable is called:

https://github.com/kgerheiser/NCEPLIBS-ip2/blob/9aab5fefb03a8dd8994ef8e03ea2107b95cd5706/tests/CMakeLists.txt#L38

GeorgeGayno-NOAA commented:
> @kgerheiser I will try your fix at [9aab5fe](https://github.com/NOAA-EMC/NCEPLIBS-ip2/commit/9aab5fefb03a8dd8994ef8e03ea2107b95cd5706).

@kgerheiser I reran my UFS_UTILS snow2mdl regression test.  It failed, but it is much closer to the baseline.  The depth data is determined from AFWA data using nearest neighbor interpolation.  That field appears to match.  The snow cover data is determined from IMS data using budget interpolation.  That field has slight differences along the snow edge.  The differences are slight enough that I think the radius of the earth is no longer the problem.  The budget interpolation has lots of options stored in the 'ipopt' array.  Can you check to see if the 'ipopt' array is being used correctly.  Here is what I use for the IMS interpolation:  https://github.com/NOAA-EMC/UFS_UTILS/blob/develop/sorc/emcsfc_snow2mdl.fd/snow2mdl.f#L250  I can help setup the snow test for you if you like.

kgerheiser commented:
Yep, I see that something is off in there.

GeorgeGayno-NOAA commented:
> Sure, it's the tests you created, but very slightly modified and combined with CMake:
> 
> https://github.com/kgerheiser/NCEPLIBS-ip2/blob/feature/combined-ip/tests/interp_mod_grib1.f90
> 
> This descriptor specifically (there's also `interp_mod_grib2.f90`):
> 
> https://github.com/kgerheiser/NCEPLIBS-ip2/blob/9aab5fefb03a8dd8994ef8e03ea2107b95cd5706/tests/interp_mod_grib1.f90#L85
> 
> Here is where the top-level executable is called:
> 
> https://github.com/kgerheiser/NCEPLIBS-ip2/blob/9aab5fefb03a8dd8994ef8e03ea2107b95cd5706/tests/CMakeLists.txt#L38

For an elliptical earth grid, the sixth element (GDS Octet 17) of the 'grd212' array should be changed from '8' to '40'.  https://www.nco.ncep.noaa.gov/pmb/docs/on388/tabled.html 
https://www.nco.ncep.noaa.gov/pmb/docs/on388/table7.html

But I would retain the 212 test and add another '212' test that is the same except for an elliptical earth.  Give it any number you want.

kgerheiser committed:
Fix subgrid budget interpolation

When the grid number is negative it means to operate on a subgrid. Gdswzd was being called on the original output grid and not the grid calculated from the subgrid.

kgerheiser commented:
I found a problem with the negative number sub-grid thing. I was calling `gdswzd` on the original output grid object (which has a negative grid number and implements `gdswzd` with an empty call), instead of on the new grid object created with `255 + grid_number`.

GeorgeGayno-NOAA commented:
> I found a problem with the negative number sub-grid thing. I was calling `gdswzd` on the original output grid object (which has a negative grid number and implements `gdswzd` with an empty call), instead of on the new grid object created with `255 + grid_number`.

Great.  I will rerun my test.

kgerheiser commented:
I was able to generate test data for a non-spherical Earth pretty easily. Now, I would also like a test for the sub grid option.  How would craft the input to do that?

GeorgeGayno-NOAA commented:
> I was able to generate test data for a non-spherical Earth pretty easily. Now, I would also like a test for the sub grid option. How would craft the input to do that?

You don't modify the input data.  Rather you interpolate the input data to a subset of the grid.  A subset is just a subsection of the input grid.   You can try a single point to start.  Allocate output_rlat/rlon, output_bitmap and output_data for one point.  Set the lat/lon to a value in the middle of North America.   Set 'no' to one.  Subtract kgds from 255.  Then call ipolates.   I think that will work.   You can expand the test to do the 'left' half of the grid.  You would need to know the lat/lons for each point on the sub grid (read them from a file?)  I use this option so I can just pass in the land points as I don't care about snow at water points.

kgerheiser committed:
Add -heap-arrays option to ifort so tests pass

kgerheiser commented:
@GeorgeGayno-NOAA could you try this again to double check me? It seems to have worked for me on Hera.

`export ip_ROOT=/home/Kyle.Gerheiser/NCEPLIBS-ip2/build/install`

GeorgeGayno-NOAA commented:
> @GeorgeGayno-NOAA could you try this again to double check me? It seems to have worked for me on Hera.
> 
> `export ip_ROOT=/home/Kyle.Gerheiser/NCEPLIBS-ip2/build/install`

Yes, it works now. Thanks.

edwardhartnett commented:
Why do we need environment variable ip_ROOT?

kgerheiser commented:
You don't, but that's one of the ways CMake finds libraries so George could build UFS_UTILS with my build, the same way as appending to `CMAKE_PREFIX_PATH`.

aerorahul commented:
> You don't, but that's one of the ways CMake finds libraries so George could build UFS_UTILS with my build, the same way as appending to `CMAKE_PREFIX_PATH`.

@edwardhartnett @kgerheiser 
One could also avoid both `pkg_ROOT` and `CMAKE_PREFIX_PATH` by passing `-Dpkg_DIR=/path/to/pkg/lib/cmake` as arguments to `cmake` at configuration time.  `/path/to/pkg/lib/cmake` should contain `pkg-config.cmake`.
Ofcourse it is easier to set `pkg_ROOT` environment variable in the module that loads the `pkg`.

kgerheiser committed:
Doxygen documentation

edwardhartnett commented:
Is this ready to merge? Lot's of changes here! ;-)

kgerheiser commented:
Well, I broke something yesterday, but otherwise it works (with UFS_UTILS, at least). I was going to finish up converting  the documentation to Doxygen. I also think it should be tested with some other application that uses ip (weather model, UPP) before it's merged.

edwardhartnett commented:
Can you save the doxygen changes for a separate PR?

kgerheiser commented:
Yes, that can be separate.

edwardhartnett commented:
@kgerheiser this can be close now, right?
